[DOC-3661][DOC-3663] Add a concept for CMEK (#13963)
cockroachdb/docs: cockroachcloud/cmek.md

---
title: Customer-Managed Encryption Keys (CMEK)
summary: Use cryptographic keys that you manage to protect data at rest in a CockroachDB Dedicated cluster.
toc: true
docs_area: manage.security
---

Customer-Managed Encryption Keys (CMEK) allow you to protect data at rest in a {{ site.data.products.dedicated }} cluster using a cryptographic key that is entirely within your control, hosted in a supported key-management system (KMS) platform. This key is called the _CMEK key_.

{{site.data.alerts.callout_info}}

> This feature is a Private Preview release, limited to customers for whom it is enabled, upon their request.

Reviewer: @abhinavg6 @mdlinville Can we add a mention of how to request it?

1. Creates the store key and encrypts it using the CMEK key.
1. Creates the data key and encrypts it using the encrypted store key.
1. Propagates the store key and data key to cluster nodes.
1. Starts the cluster.

Going forward:

1. The cluster can start only when the CMEK key is available to decrypt the store key. If the CMEK key becomes unavailable or you revoke the cluster's access to the key, the cluster can't start until access is restored.
1. Each time the store key is rotated, the new store key is also encrypted using the CMEK key.
1. Each time the data key is rotated, the new data key is encrypted using the encrypted store key.
1. Each time a node writes new data to disk, it is encrypted using the current data key. Data is read using the data key that was used to encrypt it.
{{ site.data.products.db }}
{{site.data.alerts.callout_danger}}
If the CMEK key is destroyed, the cluster's data can't be recovered or restored from a managed backup in {{ site.data.products.db }} or from a manual backup to the same cluster. It may be possible to restore a manual backup to a new cluster. 
{{site.data.alerts.end}}

> ## Backing up and restoring to a cluster with CMEK

Reviewer: ```suggestion
## Backup and restore operations on a cluster with CMEK
```
Minor suggestion here, but I prefer something like this because the current heading and its gerunds imply that the section will explain _how_ to do it, whereas this is conceptual only.

### Can the CMEK key be automatically or manually rotated?

It's not possible to rotate the CMEK key.

### Are managed backups in CockroachDB Cloud encrypted using the CMEK key?

When CMEK is enabled, managed backups in {{ site.data.products.db }} are encrypted using the CMEK key before being written to persistent storage, and the CMEK key must be available when you restore from an automatic backup. In addition, backups taken before CMEK was enabled cannot be restored to the cluster.

### Can I enable CMEK on an existing cluster?

We do not recommend enabling CMEK on an existing cluster, to prevent a situation where some data is protected with the CMEK key and other data isn't. For the same reason, we recommend that you configure CMEK after creating the cluster but before loading any data into it or giving users access to it.

This limitation also applies to adding a new region to an existing cluster. Adding a new region to a multi-region cluster that has CMEK enabled is not recommended, because the new region's data at rest will not be encrypted by the CMEK key.

> ### Can I enable CMEK from Cockroach Cloud Console?

Reviewer 2: ```suggestion
### Can I enable CMEK from CockroachDB Cloud Console?
```

It's not possible to enable CMEK from {{ site.data.products.db }} Console. Instead, you must use the [Cloud API](/docs/cockroachcloud/cloud-api.html).

### After enabling CMEK, can I disable it?

CMEK can't be disabled on a cluster after it has been enabled.

## Limitations
The CMEK feature has the following limitations:

- CMEK can be enabled only on clusters created after April 1, 2022 (AWS) or June 9, 2022 (GCP).
- To enable CMEK, you must use the [Cloud API](/docs/cockroachcloud/cloud-api.html). It's not possible to enable CMEK using the {{ site.data.products.db }} Console.

> - Adding a region to a cluster with CMEK enabled is not recommended because currently, the new region will not be protected by the CMEK key.

Reviewer 2: Again, I would like to remove this "discouragement / recommendation".

- Rotating a CMEK key is not supported.

> - Only symmetric AES-GCM software keys are supported.

Reviewer 2: I would remove this. We're not constraining anyone - it all depends on what cloud providers support. At this point it looks like we're the bottleneck.

---
title: Customer-Managed Encryption Keys (CMEK)
summary: Use cryptographic keys that you manage to protect data at rest in a CockroachDB Dedicated cluster.
toc: true
docs_area: manage.security
---

> Customer-Managed Encryption Keys (CMEK) give you more control over how a {{ site.data.products.dedicated }} cluster's data is protected at rest on the cluster's nodes. When CMEK  enabled, your cluster's data is protected by an additional cryptographic key that is entirely within your control, hosted in a supported key-management systems (KMS) platform. This key is called the _CMEK key_. You grant your cluster's service account the permission to encrypt and decrypt using the CMEK key. The CMEK key is never persisted in a cluster, and {{ site.data.products.db }} uses the KMS platform's API to encrypt and decrypt using the CMEK key. The service account never has direct access to the CMEK key materials and does not have the ability to manage or export your CMEK keys.

Author: I struggled here but I think I'm inclined to disagree. The platform is AWS, Google Cloud, or Hashicorp Vault, and you're using a KMS provided by one of them. WDYT?

(suggestions don't show up well in Reviewable, it turns out)

This article describes how CMEK works in {{ site.data.products.dedicated }} clusters. To configure CMEK, see [Managing Customer-Managed Encryption Keys (CMEK) for CockroachDB Dedicated](managing-cmek.html).

## Overview of CMEK

This section describes how CMEK works on a {{ site.data.products.dedicated }} cluster. To learn more about encryption in CockroachDB, see [Encryption At Rest](/docs/{{site.versions["stable"]}}/encryption.html). To enable CMEK, see [Managing Customer-Managed Encryption Keys (CMEK) for CockroachDB Dedicated](managing-cmek.html).

When you create a {{ site.data.products.dedicated }} cluster, two encryption keys are created automatically:

1. The _store key_ is a Key Encryption Key (KEK), and is used to protect the data key. The store key is accessed each time a cluster node is started.
1. The _data key_ is a Data Encryption Key (DEK), and is used to encrypt and decrypt cluster data before it is read from or written to a cluster node's filesystem. Each time the cluster is started or restarted, and each time a node joins the cluster, CockroachDB uses the store key to encrypt and decrypt data keys.

>   By default, data keys are rotated weekly, and CockroachDB keeps a registry of all data keys that have ever been used to write data to the cluster. When data is read from the cluster, it is first decrypted using the data key that was used to encrypt it. When data is written to a cluster, it is first encrypted using the current data key.

Author: We don't need the CMEK key because we call encrypt and decrypt using the KMS's API. It's the KMS's API that does the permission checks to make sure we have access. At cluster start-up we load the encrypted store key into memory and call the KMS's API to decrypt it. Then we can encrypt and decrypt data keys just like we did before. When we rotate the store key, we use the CMEK key to encrypt and decrypt it too, and it begins encrypting and decrypting new data keys. Does that make sense? If so, I'll try to make this section more clear.

>   By default, data keys are rotated weekly, and CockroachDB keeps a registry of all data keys that have ever been used to write data to the cluster. When data is read from the cluster, it is first decrypted using the data key that was used to encrypt it. When data is written to a cluster, it is first encrypted using the current data key.

Author: I've rewritten and de-duplicated this.

When CMEK is enabled for a cluster, the store key itself is encrypted using a symmetric key that you create and manage in a supported KMS platform. This key, which is also a KEK, is called the _CMEK key_. After the store key is encrypted, a new data key is created and encrypted using the new store key. The new store key and data key are propagated to the cluster's nodes and the previous (unencrypted) store key is deleted from the cluster's nodes.

> To grant your cluster the ability to decrypt and encrypt using the CMEK key, you use an IAM role in your key-management system.

Author: I've rewritten this significantly.

> During start-up of a CMEK-enabled cluster, CockroachDB uses the KMS platform's's API to decrypt the encrypted store key in memory using the CMEK key, and the decrypted store key is used to decrypt the data key. Data can be written to and read from the cluster only when the CMEK key is available to the cluster. In addition, the cluster's backups can be read or restored only when the CMEK key is available.

Author: Fixed during rewrite.

### Can I enable CMEK on an existing cluster?

We do not recommend enabling CMEK on an existing cluster, to prevent a situation where some data is protected with the CMEK key and other data isn't. For the same reason, we recommend that you configure CMEK after creating the cluster but before loading any data into it or giving users access to it.

This limitation also applies to adding a new region to an existing cluster. Enabling CMEK for new regions of existing clusters is not supported. Instead, create a new multi-region cluster and enable CMEK on it. For more information about using CMEK with multi-region clusters, see [Multi-region Clusters](#multi-region-clusters).

### Can I enable CMEK from Cockroach Cloud Console?

It's not possible to enable CMEK from {{ site.data.products.db }} Console. Instead, you must use the [Cloud API](/docs/cockroachcloud/cloud-api.html).

### After enabling CMEK, can I disable it?

> It's not possible to disable CMEK from a cluster where it has been enabled.

Author: I went with "CMEK can't be disabled on a cluster after it has been enabled." for simplicity.

## Limitations

- To enable CMEK, you must use the [Cloud API](/docs/cockroachcloud/cloud-api.html). It's not possible to enable CMEK using the Cloud Console.

> - The CMEK key can be stored in Amazon Web Services KMS, Google Cloud KMS, or Hashicorp Vault Secrets Managers.

Author: No, the Hashicorp product is called Hashicorp Vault Secrets Manager.
